docs/enhancements/transactional-custom-manifests.md

```diff
@@ -0,0 +1,252 @@
+
+Manifest and patch files may be marked as either "system generated" or "user generated", by default a file is considered "system generated". 
+
+To mark a file as user generated, a metadata file is created under the subtree `manifest-attributes` to match the file and path in the subtree `manifests`
+
+So if we have a file:
+`manifests/openshift/user-defined-manifest.yaml`
+
+The metadata file to mark this as user-supplied would be:
+`manifest-attributes/openshift/user-defined-manifest.yaml/user-supplied`
+
+As we are using the filesystem as a means of storing metadata about a file, this needs to be performed in two operations, one to write the file and one to write the metadata.
+
+This has a weakness; by performing two writes, we risk that the second write may not take place in the case of an S3 failure. 
+
+# Summary
+
+We would like to improve the custom manifest feature to make it more consistent, more reliable and more user friendly. By addressing the following points
+
+* Find a more reliable and "transactional" way in which the upload of manifest files may be handled.
+
+# Motivation
+
+We have had a recent bug report about an edge case
+
+  The following bug was contributed to by this issue: https://issues.redhat.com/browse/MGMT-17802
+
+In this edge case, if there is a failure in S3 with unfortunate timing then it is possible to store a manifest with inconsistent metadata.
+This is not a situation that is acceptable moving forward and we need to do something.
+
+There are some problems with the existing implementation of manifest storage when it comes to metadata
+
+* A manifest upload cannot be "transactional" under the current design, raising the risk that one part of the upload may fail and leave the rest inconsistent.
+
+### Goals
+
+- To ensure that manifests may be written atomically so that their reliability may be guaranteed.
+- To simplify the storage of manifest metadata.
+
+### Non-Goals
+
+- It is a non-goal to change the behavior of the UI for manifests.
+
+## Proposal: Use the Amazon S3 metadata API 
+
+We should change the way that we store metadata about a manifest, instead of using an awkward "directory based" approach to metadata, we should instead use the built in facilities of S3 which has native Metadata support for objects.
+This should be achievable in a single write and should mitigate the transactionality issues.
+
+### Object store API changes required:
+
+	#### Uploads
+
+	S3Wrapper API:
+	Calls to Upload, UploadFile and UploadStream will need to be updated to support metadata for an object
+	https://github.com/openshift/assisted-service/blob/master/pkg/s3wrapper/client.go#L42-L43
+	
+	##### Filesystem implementation:
+	
+	Implementation changes to the filesystem S3Wrapper client will be required in order to implement metadata support.
+	We should attempt to keep this atomic by using XATTRs if supported.
```

Comment:
Take into account that setting XATTRs requires a separate `setxattr` system call. So you will need first to create the file and then set the extended attributes. One of those two operations may fail, or never be attempted (if you crash in between) so it won't give you the transactional behavior that you are looking for.

I think that our filesystem implementation creates a temporary file and them, when it is completely written, it moves it to its definitive location. So to get that transactional behavior you will need to make sure that the extended attribute is set on the temporary file before moving it.

Reply:
This is a very good point, I will be sure to keep this in mind when working on this.
One thing we do get rid of is distributed + non transactional, easier to deal with rollbacks and issues such as this when the filesystem is local (the original file can be deleted before returning an error)

Not an option in remote filesystems such as S3 of course.

Reply:
Yes I thought of this, but we can get around it by creating the file with the xattr, then renaming it.  The rename operation is transactional.
Though now that I'm reading this document, why not just have separate directories for user-generated manifests?

Instead of
manifests/openshift/user-defined-manifest.yaml
Upload it to
manifests/openshift.userdefined/manifest.yaml

Then you don't need any additional file or attribute, because anything under there is user-defined.

Reply:
I actually think that using the AWS metadata is the right approach

I think it's the right approach because using a folder to "hack" some metadata is not a scalable approach and certainly not pleasant to work with or write tests for

using AWS metadata + Xattr (for the local implementation) makes more sense to me as it can be extended in future with minimal effort.